File: 2024/16.py
import time, heapq

from utils import readInput

# TOGLIERE PARTE 1 e FARE TUTTO CON LA DUE PROVARE AD USARE NUMERI COMPLESSI CON HEAPQ

# On to the Python trick of the day: more and more people are starting to use complex numbers for grid puzzles, and they might have hit a roadblock when using them in a priority queue.
# Suppose you have a queue of (score, position) tuples. As long as the scores are unique, they can fully determine the order of the queue. But when there are duplicate scores (which can easily happen today), Python wants to sort on the second element, position.
# Since complex numbers can't be sorted (1+9j isn't necessarily "less" than 2+2j), Python throws an error:
# TypeError: '<' not supported between instances of 'complex' and 'complex'
# There are a few ways to mitigate this:
# write your own complex number class, inheriting from the built-in complex but redefining less-than (u/xelf did this here),
# store the number as a string, and "re-hydrate" it to complex upon retrieval (u/atreju3647 did this here),
# store the real and imaginary parts separately, and combine them upon retrieval (u/TiCoinCoin did this here),
# when inserting to the priority queue, add a "tie-breaker" to your tuple. So (score, position) becomes (score, t, position), where t is a unique value. This can be a random number, or an ever incrementing value.
# Here's a simple example:
# from heapq import heappush, heappop
# Q = [(1, i:=0, complex(1))]
# for x in [2, 3, 4]:
#     heappush(Q, (x, i := i+1, complex(x,x)))
# When extracting values from the queue, just ignore the tie-breaker:
# x, _, y = heappop(Q)
# If anyone has questions, suggestions or other solutions, feel free to let me know!

DIRECTIONS = [(0, 1), (1, 0), (0, -1), (-1, 0)]

def loadInput(filename: str):
    lines = readInput(filename)

    pitch = {}
    start = None
    target = None
    for y in range(len(lines)):
        for x in range(len(lines[0])):
            c = (x, y)
            if lines[y][x] == "#":
                pitch[c] = 1
            elif lines[y][x] == "S":
                pitch[c] = 0
                start = c
            elif lines[y][x] == "E":
                target = c
                pitch[c] = 0
            elif lines[y][x] == ".":
                pitch[c] = 0
    #draw(pitch, start, target)                       
    return pitch, start, target

def is_valid(grid, c):
    return c in grid and grid[c] != 1

def get_neighbors(grid, c, direction):
    neighbors = []
    for idc, dc in enumerate(DIRECTIONS):
        if dc == (-DIRECTIONS[direction][0], -DIRECTIONS[direction][1]):
            continue
        nc = (c[0] + dc[0], c[1] + dc[1])
        if is_valid(grid, nc):
            if dc == DIRECTIONS[direction]: 
                neighbors.append((nc, direction, 1))
            else:
                neighbors.append((nc, idc, 1001))
    return neighbors

def solve(grid, start, target):
    q = [(0, start, 1)]
    visited = set()
    predecessors = {}

    while q:
        cost, c, direction = heapq.heappop(q)
        if (c, direction) in visited:
            continue
        visited.add((c, direction))

        if c == target:
            path = []
            current = (c, direction)
            while current:
                path.append((current[0],current[1]))
                current = predecessors.get(current)
            return cost, path[::-1]

        for nc, nd, move_cost in get_neighbors(grid, c, direction):
            neighbor = (nc, nd)
            if neighbor not in visited:
                new_cost = cost + move_cost
                heapq.heappush(q, (new_cost, nc, nd))
                predecessors[neighbor] = (c, direction)
    return -1, None

def part1(pitch, start, target):
    cost, path = solve(pitch, start, target)
    #draw(pitch, path)
    print (f"🎄 Part 1: {cost}", end='')

def solve2(grid, start, target):
    DIRECTIONS = set([(0, 1), (1, 0), (0, -1), (-1, 0)])
    q = [(0, start, (1, 0), set())]  # score, position, direction, visited tiles
    seen_states = {
        pos: {direction: {"score": float("inf"), "tiles": set()} for direction in DIRECTIONS}
        for pos in grid.keys()
    }
    while q:
        score, coord, direction, tiles = heapq.heappop(q)
        allowed = DIRECTIONS - {(-direction[0], -direction[1])}  # avoid reversing direction
        for dc in allowed:
            new_coord = (coord[0] + dc[0], coord[1] + dc[1])
            if not is_valid(grid, new_coord):
                continue
            new_score = score + 1 + (1000 if dc != direction else 0)
            state = seen_states[new_coord][dc]
            if new_score > state["score"]:
                continue
            new_tiles = tiles | {new_coord}
            state["score"] = new_score
            if new_score < state["score"]:
                state["tiles"] = new_tiles
            elif new_score == state["score"]:
                state["tiles"] |= new_tiles        
            heapq.heappush(q, (new_score, new_coord, dc, state["tiles"]))
    return seen_states
    
def part2(pitch, start, target):
    visited = solve2(pitch, start, target)
    min_direction = min(visited[target].values(), key=lambda d: d["score"])
    #draw(pitch, min_direction['tiles'])
    print (f"🎄🎅 Part 2: {len(min_direction["tiles"]) + 1}", end='')

def draw(pitch, sits):
    xmax = int(max([c[0] for c in pitch]))+1
    ymax = int(max([c[1] for c in pitch]))+1
    print ( xmax, ymax)
    #points = {c[0]:c[1] for c in sits}
    dirs = ['v','>','^','<']
    for y in range(ymax):
        for x in range(xmax):
            c = (x, y)
            if c in sits:
                #print (dirs[points[c]], end='')
                print ("O", end='')
            elif pitch[c] == 1:
                print ("#", end='')
            elif pitch[c] == 0:
                print (".", end='')
        print ("")

if __name__ == '__main__':
    title = "Day 16: Reindeer Maze"
    sub = "❄ "*(len(title)//2+2)

    print()
    print(f" {title} ")
    print(sub)
    
    inputs = loadInput("input_16.txt")
    
    t0 = time.time()
    part1(*inputs)
    print (" - {:.5f}".format(time.time()-t0))
    
    t0 = time.time()
    part2(*inputs)
    print (" - {:.5f}".format(time.time()-t0))

File: 2024/16.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <queue>
#include <set>
#include <map>
#include <limits>
#include <chrono>

using namespace std;
using namespace std::chrono;

struct Point {
    int x, y;
    bool operator==(const Point& other) const {
        return x == other.x && y == other.y;
    }
    bool operator<(const Point& other) const {
        if (y != other.y) return y < other.y;
        return x < other.x;
    }
};

struct State {
    Point p;
    int dir;
    int cost;

    bool operator>(const State& other) const {
        return cost > other.cost;
    }
};

struct State2 {
    Point p;
    Point dir;
    int score;
    set<Point> tiles;
};

const vector<Point> DIRECTIONS = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

map<Point, int> loadInput(const string& filename, Point& start, Point& target) {
    ifstream file(filename);
    if (!file.is_open()) {
        cerr << "Error opening file: " << filename << endl;
        exit(1);
    }

    map<Point, int> pitch;
    vector<string> lines;
    string line;
    while (getline(file, line)) {
        lines.push_back(line);
    }

    for (int y = 0; y < lines.size(); ++y) {
        for (int x = 0; x < lines[0].size(); ++x) {
            Point c = {x, y};
            if (lines[y][x] == '#') {
                pitch[c] = 1;
            } else if (lines[y][x] == 'S') {
                pitch[c] = 0;
                start = c;
            } else if (lines[y][x] == 'E') {
                target = c;
                pitch[c] = 0;
            } else if (lines[y][x] == '.') {
                pitch[c] = 0;
            }
        }
    }
    return pitch;
}

bool isValid(const map<Point, int>& grid, const Point& c) {
    return grid.count(c) > 0 && grid.at(c) != 1;
}

vector<pair<State, int>> getNeighbors(const map<Point, int>& grid, const Point& c, int direction) {
    vector<pair<State, int>> neighbors;
    for (int idc = 0; idc < DIRECTIONS.size(); ++idc) {
        Point dc = DIRECTIONS[idc];
        if (dc.x == -DIRECTIONS[direction].x && dc.y == -DIRECTIONS[direction].y) {
            continue;
        }
        Point nc = {c.x + dc.x, c.y + dc.y};
        if (isValid(grid, nc)) {
            int moveCost = (dc.x == DIRECTIONS[direction].x && dc.y == DIRECTIONS[direction].y) ? 1 : 1001;
            neighbors.push_back({{nc, idc, 0}, moveCost});
        }
    }
    return neighbors;
}

pair<int, vector<pair<Point, int>>> part1(const map<Point, int>& grid, const Point& start, const Point& target) {
    priority_queue<State, vector<State>, greater<State>> q;
    q.push({start, 1, 0});
    set<pair<Point, int>> visited;
    map<pair<Point, int>, pair<Point, int>> predecessors;

    while (!q.empty()) {
        State current_state = q.top();
        q.pop();
        Point c = current_state.p;
        int direction = current_state.dir;
        int cost = current_state.cost;

        if (visited.count({c, direction})) continue;
        visited.insert({c, direction});

        if (c == target) {
            vector<pair<Point, int>> path;
            pair<Point, int> current = {c, direction};
            while (predecessors.count(current)) {
                path.push_back(current);
                current = predecessors.at(current);
            }
            cout << "🎄 Part 1: " << cost;
            return {cost, vector<pair<Point,int>>(path.rbegin(), path.rend())};
        }

        for (auto& neighbor_data : getNeighbors(grid, c, direction)) {
            Point nc = neighbor_data.first.p;
            int nd = neighbor_data.first.dir;
            int move_cost = neighbor_data.second;

            if (!visited.count({nc, nd})) {
                int new_cost = cost + move_cost;
                q.push({nc, nd, new_cost});
                predecessors[{nc, nd}] = {c, direction};
            }
        }
    }
    return {-1, {}};
}

void part2(const map<Point, int>& pitch, const Point& start, const Point& target) {
    map<Point, map<Point, map<string, set<Point>>>> seen_states;
    for (const auto& [pos, val] : pitch) {
        for (Point dir : DIRECTIONS) {
            seen_states[pos][dir]["score"] = {};
            seen_states[pos][dir]["tiles"] = {};
        }
    }

    priority_queue<State2, vector<State2>, function<bool(const State2&, const State2&)>> q([](const State2& a, const State2& b) {
        return a.score > b.score;
    });

    q.push({start, {1, 0}, 0, {}});

    while (!q.empty()) {
        State2 current = q.top();
        q.pop();
        Point coord = current.p;
        Point direction = current.dir;
        set<Point> tiles = current.tiles;
        int score = current.score;

        set<Point> allowed = {{0,1}, {1,0}, {0,-1}, {-1,0}};
        for(Point p : DIRECTIONS){
            if(p.x == -direction.x && p.y == -direction.y){
                allowed.erase(p);
                break;
            }
        }

        for (Point dc : allowed) {
            Point new_coord = {coord.x + dc.x, coord.y + dc.y};
            if (!isValid(pitch, new_coord)) {
                continue;
            }
            int new_score = score + 1 + (dc.x != direction.x || dc.y != direction.y ? 1000 : 0);

            if (seen_states.count(new_coord) && seen_states[new_coord].count(dc)) {
                if (!seen_states[new_coord][dc]["score"].empty() && new_score > *seen_states[new_coord][dc]["score"].begin()) {
                    continue;
                }
            } else {
                seen_states[new_coord][dc]["score"].insert(numeric_limits<int>::max());
            }


            set<Point> new_tiles = tiles;
            new_tiles.insert(new_coord);

             seen_states[new_coord][dc]["score"].clear();
            seen_states[new_coord][dc]["score"].insert(new_score);
            seen_states[new_coord][dc]["tiles"] = new_tiles;

            q.push({new_coord, dc, new_score, new_tiles});
        }
    }

    int min_score = numeric_limits<int>::max();
    set<Point> min_tiles;
    for (const auto& [dir, data] : seen_states[target]) {
        if (!data["score"].empty() && *data["score"].begin() < min_score) {
            min_score = *data["score"].begin();
            min_tiles = data["tiles"];
        }
    }
    cout << "🎄🎅 Part 2: " << min_tiles.size() + 1 << endl;
}

void draw(const map<Point, int>& pitch, const set<Point>& sits) {
    int xmax = 0, ymax = 0;
    for (const auto& [c, v] : pitch) {
        xmax = max(xmax, c.x);
        ymax = max(ymax, c.y);
    }
    xmax++; ymax++;

    for (int y = 0; y < ymax; ++y) {
        for (int x = 0; x < xmax; ++x) {
            Point c = {x, y};
            if (sits.count(c)) {
                cout << "O";
            } else if (pitch.count(c) && pitch.at(c) == 1) {
                cout << "#";
            } else {
                cout << ".";
            }
        }
        cout << endl;
    }
}

int main() {
    string title = "Day 16: Reindeer Maze";
    string sub = string(title.length() + 2, '-');

    cout << endl;
    cout << " " << title << " " << endl;
    cout << sub << endl;

    Point start, target;
    map<Point, int> grid = loadInput("input_16.txt", start, target);

    auto t0 = std::chrono::high_resolution_clock::now();
    part1(grid, start, target);
    auto t1 = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(t1 - t0);
    std::cout << " - " << duration.count() << "\u00B5s" << std::endl; 

    auto t0 = std::chrono::high_resolution_clock::now();
    part2(grid, start, target);
    auto t1 = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(t1 - t0);
    std::cout << " - " << duration.count() << "\u00B5s" << std::endl; 

//   start = std::chrono::high_resolution_clock::now();
//   part2(memory);
//   end = std::chrono::high_resolution_clock::now();
//   duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
//   std::cout << " - " << duration.count() << "\u00B5s" << std::endl; 

    return 0;
}

// void part2(const map<Point, int>& pitch, const Point& start, const Point& target) {
//     map<Point, map<Point, map<string, set<Point>>>> seen_states;
//     for (const auto& [pos, val] : pitch) {
//         seen_states[pos][{0,0}]["score"] = set<Point>();
//         seen_states[pos][{0,0}]["tiles"] = set<Point>();

//         for(Point dir : DIRECTIONS){
//             seen_states[pos][dir]["score"] = set<Point>();
//             seen_states[pos][dir]["tiles"] = set<Point>();
//         }
//     }

//     priority_queue<State2, vector<State2>, function<bool(const State2&, const State2&)>> q([](const State2& a, const State2& b) {
//         return a.score > b.score;
//     });


//     q.push({start, {1, 0}, 0, {}});

//     while (!q.empty()) {
//         State2 current = q.top();
//         q.pop();
//         Point coord = current.p;
//         Point direction = current.dir;
//         set<Point> tiles = current.tiles;
//         int score = current.score;

//         set<Point> allowed = {{0,1}, {1,0}, {0,-1}, {-1,0}};
//         for(Point p : allowed){
//             if(p.x == -direction.x && p.y == -direction.y){
//                 allowed.erase(p);
//                 break;
//             }
//         }

//         for (Point dc : allowed) {
//             Point new_coord = {coord.x + dc.x, coord.y + dc.y};
//             if (!isValid(pitch, new_coord)) {
//                 continue;
//             }
//             int new_score = score + 1 + (dc.x != direction.x || dc.y != direction.y ? 1000 : 0);

//             if(seen_states.count(new_coord) == 0 || seen_states[new_coord].co#include <chrono>

// void part1(string& memory) {
//   const regex pattern("mul\\((\\d{1,3}),(\\d{1,3})\\)");//|do\\(\\)|don't\\(\\)");
//   int tot = 0;
//   sregex_iterator iter(memory.begin(), memory.end(), pattern);
//   sregex_iterator end_iter;
  
//   while (iter != end_iter) {
//     smatch match = *iter;
    
//     tot += stoi(match[1])*stoi(match[2]);
//     ++iter;
//   }

//   std::cout << "🎄   Part 1: " << tot;
// }

// void part2(string& memory) {

//   const regex pattern("mul\\((\\d{1,3}),(\\d{1,3})\\)|do\\(\\)|don't\\(\\)");
//   int tot = 0;
//   bool enabled = true;
//   sregex_iterator iter(memory.begin(), memory.end(), pattern);
//   sregex_iterator end_iter;
  
//   while (iter != end_iter) {
//     smatch match = *iter;
//     if (match.str() == "do()")
//       enabled = true;
//     else if (match.str() == "don't()")
//       enabled = false;
//     else {
//       if (enabled)
// 	tot += stoi(match[1])*stoi(match[2]);
//     }
//     ++iter;
//   }

//   std::cout << "🎄🎅 Part 2: " << tot;
// }

// int main() {
//   ifstream file("input_3.txt");
  
//   if (!file.is_open()) {
//     cerr << "Error opening file!" << endl;
//     return 1;
//   }
  
//   string memory;
//   string line;
//   while (getline(file, line)) {
//     memory += line;  
//   }

//   file.close();

//   std::cout << "Day 3: Mull It Over" << std::endl;


//   return 0;
// }
